### ISSUE TITLE
raise psr/http-message to version 2

### ISSUE THREAD

**Issue by pfpro** [mpdf/mpdf#1878]
### Please describe the new functionality as best as you can.

good evening. is is possible to raise "psr/http-message" in composer.json to version 2?

**Comment by finwe** [comment]
It is, currently only the [same way the psr/log 3 support was done.](https://github.com/mpdf/mpdf/pull/1857)

**Comment by Ph0tonic** [comment]
Hi,
I was willing to help push forward this feature when I realised all the work needed just to support PHP 5.6. I have read your previous discussions here https://github.com/mpdf/mpdf/issues/1263 (3 years ago), and the main argument was to support the wider range of PHP versions. By dropping php 5.6 only (end of maintenance 6 years ago) and keeping php ^7.0, developing this feature would be really simple as php 7 support return types. Here is the stats usage from Packagist, which I believe represents quite well stats of active development projects and php 5.6 represents less than 1%.
- https://packagist.org/php-statistics

From this perspective, I personally think that it might be time to move forward.

**Comment by finwe** [comment]
If it were as easy as dropping support of PHP version, it would be done already. https://github.com/mpdf/mpdf/issues/1637

**Comment by Ph0tonic** [comment]
Sorry, if I missed something but here is what I did when PHP 5.6 is dropped (see #1884).

**Comment by finwe** [comment]
Thanks for your effort but not ATM.

**Comment by NigelCunningham** [comment]
Starting work on a PR here: https://github.com/NigelCunningham/mpdf/tree/http_message_2_support

**Comment by michaelw85** [comment]
It would be great if packages started following the [PHP release](https://www.php.net/supported-versions.php) cadence and stop encouraging using EOL versions.

For anyone who runs into fatal errors because other packages do install a modern PSR standard, here's a git patch for mpdf:

<details>
  <summary>PSR Patch</summary>

  ```
  diff --git a/mpdf/src/Http/Request.php b/mpdf/src/Http/Request.php
  index 6027daa..02f364d 100644
  --- a/mpdf/src/Http/Request.php
  +++ b/mpdf/src/Http/Request.php
  @@ -4,13 +4,15 @@ namespace Mpdf\Http;
   
   use Psr\Http\Message\StreamInterface;
   use Psr\Http\Message\UriInterface;
  +use Psr\Http\Message\RequestInterface;
  +use Psr\Http\Message\MessageInterface;
   
   /**
    * PSR-7 URI implementation ported from nyholm/psr7 and adapted for PHP 5.6
    *
    * @link https://github.com/Nyholm/psr7/blob/master/src/Uri.php
    */
  -class Request implements \Psr\Http\Message\RequestInterface
  +class Request implements RequestInterface
   {
   
 	  /** @var string */
  @@ -66,7 +68,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  }
 	  }
   
  -	public function getRequestTarget()
  +	public function getRequestTarget(): string
 	  {
 		  if ($this->requestTarget !== null) {
 			  return $this->requestTarget;
  @@ -83,7 +85,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $target;
 	  }
   
  -	public function withRequestTarget($requestTarget)
  +	public function withRequestTarget(string $requestTarget): RequestInterface
 	  {
 		  if (preg_match('#\s#', $requestTarget)) {
 			  throw new \InvalidArgumentException('Invalid request target provided; cannot contain whitespace');
  @@ -95,12 +97,12 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function getMethod()
  +	public function getMethod(): string
 	  {
 		  return $this->method;
 	  }
   
  -	public function withMethod($method)
  +	public function withMethod(string $method): RequestInterface
 	  {
 		  $new = clone $this;
 		  $new->method = $method;
  @@ -108,12 +110,12 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function getUri()
  +	public function getUri(): UriInterface
 	  {
 		  return $this->uri;
 	  }
   
  -	public function withUri(UriInterface $uri, $preserveHost = false)
  +	public function withUri(UriInterface $uri, bool $preserveHost = false): RequestInterface
 	  {
 		  if ($uri === $this->uri) {
 			  return $this;
  @@ -152,12 +154,12 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  $this->headers = [$header => [$host]] + $this->headers;
 	  }
   
  -	public function getProtocolVersion()
  +	public function getProtocolVersion(): string
 	  {
 		  return $this->protocol;
 	  }
   
  -	public function withProtocolVersion($version)
  +	public function withProtocolVersion(string $version): MessageInterface
 	  {
 		  if ($this->protocol === $version) {
 			  return $this;
  @@ -169,17 +171,17 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function getHeaders()
  +	public function getHeaders(): array
 	  {
 		  return $this->headers;
 	  }
   
  -	public function hasHeader($header)
  +	public function hasHeader(string $header): bool
 	  {
 		  return isset($this->headerNames[strtolower($header)]);
 	  }
   
  -	public function getHeader($header)
  +	public function getHeader(string $header): array
 	  {
 		  $header = strtolower($header);
   
  @@ -192,12 +194,12 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $this->headers[$header];
 	  }
   
  -	public function getHeaderLine($header)
  +	public function getHeaderLine(string $name): string
 	  {
 		  return implode(', ', $this->getHeader($header));
 	  }
   
  -	public function withHeader($header, $value)
  +	public function withHeader(string $name, $value): MessageInterface
 	  {
 		  if (!is_array($value)) {
 			  $value = [$value];
  @@ -216,7 +218,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function withAddedHeader($header, $value)
  +	public function withAddedHeader(string $name, $value): MessageInterface
 	  {
 		  if (!is_array($value)) {
 			  $value = [$value];
  @@ -237,7 +239,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function withoutHeader($header)
  +	public function withoutHeader(string $name): MessageInterface
 	  {
 		  $normalized = strtolower($header);
   
  @@ -253,7 +255,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $new;
 	  }
   
  -	public function getBody()
  +	public function getBody(): StreamInterface
 	  {
 		  if (!$this->stream) {
 			  $this->stream = Stream::create('');
  @@ -263,7 +265,7 @@ class Request implements \Psr\Http\Message\RequestInterface
 		  return $this->stream;
 	  }
   
  -	public function withBody(StreamInterface $body)
  +	public function withBody(StreamInterface $body): MessageInterface
 	  {
 		  if ($body === $this->stream) {
 			  return $this;
  diff --git a/mpdf/src/Http/Response.php b/mpdf/src/Http/Response.php
  index d562079..43b6887 100644
  --- a/mpdf/src/Http/Response.php
  +++ b/mpdf/src/Http/Response.php
  @@ -3,13 +3,15 @@
   namespace Mpdf\Http;
   
   use Psr\Http\Message\StreamInterface;
  +use Psr\Http\Message\ResponseInterface;
  +use Psr\Http\Message\MessageInterface;
   
   /**
    * PSR-7 URI implementation ported from nyholm/psr7 and adapted for PHP 5.6
    *
    * @link https://github.com/Nyholm/psr7/blob/master/src/Uri.php
    */
  -class Response implements \Psr\Http\Message\ResponseInterface
  +class Response implements ResponseInterface
   {
   
 	  /** @var array Map of standard HTTP status code/reason phrases */
  @@ -64,17 +66,17 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  $this->protocol = $version;
 	  }
   
  -	public function getStatusCode()
  +	public function getStatusCode(): int
 	  {
 		  return $this->statusCode;
 	  }
   
  -	public function getReasonPhrase()
  +	public function getReasonPhrase(): string
 	  {
 		  return $this->reasonPhrase;
 	  }
   
  -	public function withStatus($code, $reasonPhrase = '')
  +	public function withStatus(int $code, string $reasonPhrase = ''): ResponseInterface
 	  {
 		  if (!\is_int($code) && !\is_string($code)) {
 			  throw new \InvalidArgumentException('Status code has to be an integer');
  @@ -95,12 +97,12 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $new;
 	  }
   
  -	public function getProtocolVersion()
  +	public function getProtocolVersion(): string
 	  {
 		  return $this->protocol;
 	  }
   
  -	public function withProtocolVersion($version)
  +	public function withProtocolVersion(string $version): MessageInterface
 	  {
 		  if ($this->protocol === $version) {
 			  return $this;
  @@ -112,17 +114,17 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $new;
 	  }
   
  -	public function getHeaders()
  +	public function getHeaders(): array
 	  {
 		  return $this->headers;
 	  }
   
  -	public function hasHeader($header)
  +	public function hasHeader(string $header): bool
 	  {
 		  return isset($this->headerNames[strtolower($header)]);
 	  }
   
  -	public function getHeader($header)
  +	public function getHeader(string $header): array
 	  {
 		  $header = strtolower($header);
   
  @@ -135,12 +137,12 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $this->headers[$header];
 	  }
   
  -	public function getHeaderLine($header)
  +	public function getHeaderLine(string $header): string
 	  {
 		  return implode(', ', $this->getHeader($header));
 	  }
   
  -	public function withHeader($header, $value)
  +	public function withHeader(string $header, $value): MessageInterface
 	  {
 		  if (!is_array($value)) {
 			  $value = [$value];
  @@ -159,7 +161,7 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $new;
 	  }
   
  -	public function withAddedHeader($header, $value)
  +	public function withAddedHeader(string $header, $value): MessageInterface
 	  {
 		  if (!is_array($value)) {
 			  $value = [$value];
  @@ -180,7 +182,7 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $new;
 	  }
   
  -	public function withoutHeader($header)
  +	public function withoutHeader(string $header): MessageInterface
 	  {
 		  $normalized = strtolower($header);
   
  @@ -196,7 +198,7 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $new;
 	  }
   
  -	public function getBody()
  +	public function getBody(): StreamInterface
 	  {
 		  if (!$this->stream) {
 			  $this->stream = Stream::create('');
  @@ -205,7 +207,7 @@ class Response implements \Psr\Http\Message\ResponseInterface
 		  return $this->stream;
 	  }
   
  -	public function withBody(StreamInterface $body)
  +	public function withBody(StreamInterface $body): MessageInterface
 	  {
 		  if ($body === $this->stream) {
 			  return $this;
  diff --git a/mpdf/src/Http/Uri.php b/mpdf/src/Http/Uri.php
  index 553ee44..5e4203c 100644
  --- a/mpdf/src/Http/Uri.php
  +++ b/mpdf/src/Http/Uri.php
  @@ -2,12 +2,14 @@
   
   namespace Mpdf\Http;
   
  +use Psr\Http\Message\UriInterface;
  +
   /**
    * PSR-7 URI implementation ported from nyholm/psr7 and adapted for PHP 5.6
    *
    * @link https://github.com/Nyholm/psr7/blob/master/src/Uri.php
    */
  -final class Uri implements \Psr\Http\Message\UriInterface
  +final class Uri implements UriInterface
   {
 	  private static $schemes = ['http' => 80, 'https' => 443];
   
  @@ -62,12 +64,12 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return self::createUriString($this->scheme, $this->getAuthority(), $this->path, $this->query, $this->fragment);
 	  }
   
  -	public function getScheme()
  +	public function getScheme(): string
 	  {
 		  return $this->scheme;
 	  }
   
  -	public function getAuthority()
  +	public function getAuthority(): string
 	  {
 		  if ('' === $this->host) {
 			  return '';
  @@ -85,37 +87,37 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $authority;
 	  }
   
  -	public function getUserInfo()
  +	public function getUserInfo(): string
 	  {
 		  return $this->userInfo;
 	  }
   
  -	public function getHost()
  +	public function getHost(): string
 	  {
 		  return $this->host;
 	  }
   
  -	public function getPort()
  +	public function getPort(): int|null
 	  {
 		  return $this->port;
 	  }
   
  -	public function getPath()
  +	public function getPath(): string
 	  {
 		  return $this->path;
 	  }
   
  -	public function getQuery()
  +	public function getQuery(): string
 	  {
 		  return $this->query;
 	  }
   
  -	public function getFragment()
  +	public function getFragment(): string
 	  {
 		  return $this->fragment;
 	  }
   
  -	public function withScheme($scheme)
  +	public function withScheme(string $scheme): UriInterface
 	  {
 		  if (!\is_string($scheme)) {
 			  throw new \InvalidArgumentException('Scheme must be a string');
  @@ -132,7 +134,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withUserInfo($user, $password = null)
  +	public function withUserInfo(string $user, ?string $password = null): UriInterface
 	  {
 		  $info = $user;
 		  if (null !== $password && '' !== $password) {
  @@ -149,7 +151,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withHost($host)
  +	public function withHost(string $host): UriInterface
 	  {
 		  if (!\is_string($host)) {
 			  throw new \InvalidArgumentException('Host must be a string');
  @@ -165,7 +167,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withPort($port)
  +	public function withPort(?int $port): UriInterface
 	  {
 		  if ($this->port === $port = $this->filterPort($port)) {
 			  return $this;
  @@ -177,7 +179,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withPath($path)
  +	public function withPath(string $path): UriInterface
 	  {
 		  if ($this->path === $path = $this->filterPath($path)) {
 			  return $this;
  @@ -189,7 +191,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withQuery($query)
  +	public function withQuery(string $query): UriInterface
 	  {
 		  if ($this->query === $query = $this->filterQueryAndFragment($query)) {
 			  return $this;
  @@ -201,7 +203,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 		  return $new;
 	  }
   
  -	public function withFragment($fragment)
  +	public function withFragment(string $fragment): UriInterface
 	  {
 		  if ($this->fragment === $fragment = $this->filterQueryAndFragment($fragment)) {
 			  return $this;
  @@ -216,7 +218,7 @@ final class Uri implements \Psr\Http\Message\UriInterface
 	  /**
 	   * Create a URI string from its various parts.
 	   */
  -	private static function createUriString($scheme, $authority, $path, $query, $fragment)
  +	private static function createUriString(string $scheme, string $authority, string $path, string $query, string $fragment): string
 	  {
 		  $uri = '';
 		  if ('' !== $scheme) {

  ```
</details>

**Comment by season886** [comment]
php 8.2.8 + laravel 10 report 
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires mpdf/mpdf 8.1.6 -> satisfiable by mpdf/mpdf[v8.1.6].
    - mpdf/mpdf v8.1.6 requires psr/http-message ^1.0 -> found psr/http-message[1.0, 1.0.1, 1.1] but the package is fixed to 2.0 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.

Use the option --with-all-dependencies (-W) to allow upgrades, downgrades and removals for packages currently locked to specific versions.

Installation failed, reverting ./composer.json and ./composer.lock to their original content.
```

**Comment by finwe** [comment]
Resolved in #1907

**Comment by pfpro** [comment]
thank you guys!
